cirq-core/cirq/transformers/transformer_primitives.py

```diff
@@ -304,54 +342,203 @@ class _MergedCircuit:
     ckey_indexes: dict[cirq.MeasurementKey, list[int]] = dataclasses.field(
         default_factory=lambda: defaultdict(lambda: [-1])
     )
-    ops_by_index: list[dict[cirq.Operation, int]] = dataclasses.field(default_factory=list)
+    components_by_index: list[dict[Component, int]] = dataclasses.field(default_factory=list)
```

Comment:
Nit - if dictionary values are unused, it is perhaps better to declare it as `dict[Component, NoneType]` and use `None` instead of `0` to add to dictionary.